src/sema/assembly/functions.rs

```diff
@@ -1,23 +1,279 @@
+        ns.diagnostics.push(Diagnostic {
+            level: Level::Error,
+            ty: ErrorType::DeclarationError,
+            pos: func_def.id.loc,
+            message: format!("function '{}' is already defined", func_def.id.name),
+            notes: vec![Note {
+                pos: defined_func.id.loc,
+                message: "found definition here".to_string(),
+            }],
+        });
+        return;
+    } else if parse_builtin_keyword(&func_def.id.name).is_some()
+        || assembly_unsupported_builtin(&func_def.id.name)
+    {
+        ns.diagnostics.push(Diagnostic::error(
+            func_def.loc,
+            format!(
+                "function '{}' is a built-in function and cannot be redefined",
+                func_def.id.name
+            ),
+        ));
+        return;
+    } else if func_def.id.name.starts_with("verbatim") {
+        ns.diagnostics.push(Diagnostic::error(
+            func_def.id.loc,
+            "the prefix 'verbatim' is reserved for verbatim functions".to_string(),
+        ));
+        return;
+    }
+
+    let params = process_parameters(&func_def.params, ns);
+    let returns = process_parameters(&func_def.returns, ns);
+
+    if let Some(diagnostic) = functions_table.add_function_header(&func_def.id, params, returns) {
+        ns.diagnostics.push(diagnostic);
+    }
+}
+
+/// Semantic analysis of function definitions
+pub(crate) fn process_function_definition(
+    func_def: &pt::AssemblyFunctionDefinition,
+    functions_table: &mut FunctionsTable,
+    context: &ExprContext,
+    ns: &mut Namespace,
+) -> Result<AssemblyFunction, ()> {
+    let mut symtable = Symtable::new();
+    let mut local_ctx = context.clone();
+    local_ctx.yul_function = true;
+    functions_table.new_scope();
+
+    let (params, returns) = functions_table.get_params_and_returns(&func_def.id.name);
+
+    for item in &*params {
+        if symtable
+            .exclusive_add(
+                &item.id,
+                item.ty.clone(),
+                ns,
+                VariableInitializer::Assembly(true),
+                VariableUsage::AssemblyLocalVariable,
+                None,
+            )
+            .is_none()
+        {
+            ns.diagnostics.push(Diagnostic::error(
+                item.id.loc,
+                format!("variable name '{}' already taken", item.id.name),
+            ));
+            return Err(());
+        }
+    }
+
+    for item in &*returns {
+        if symtable
+            .exclusive_add(
+                &item.id,
+                item.ty.clone(),
+                ns,
+                VariableInitializer::Assembly(false),
+                VariableUsage::AssemblyLocalVariable,
+                None,
+            )
+            .is_none()
+        {
+            ns.diagnostics.push(Diagnostic::error(
+                item.id.loc,
+                format!("variable name '{}' already taken", item.id.name),
```

Comment:
This is not a good error message. Also what is the location of the previous definition?